// better-auth/dist/oauth2/index.mjs
export { a as createAuthorizationCodeRequest, c as createAuthorizationURL, d as createRefreshAccessTokenRequest, i as decryptOAuthToken, h as encodeOAuthParameter, e as generateCodeChallenge, g as generateState, f as getOAuth2Tokens, j as handleOAuthUserInfo, p as parseState, r as refreshAccessToken, s as setTokenUtil, v as validateAuthorizationCode, b as validateToken } from '../shared/better-auth.C3-_8m-g.mjs';
import 'zod';
import '../shared/better-auth.C5vDERgF.mjs';
import 'better-call';
import '../shared/better-auth.CW6D9eSx.mjs';
import '../shared/better-auth.DiMXeqeq.mjs';
import '../shared/better-auth.DdzSJf-n.mjs';
import '../shared/better-auth.CiuwFiHM.mjs';
import '@better-auth/utils/base64';
import '@better-auth/utils/hmac';
import '../shared/better-auth.BTrSrKsi.mjs';
import '../shared/better-auth.DgGir396.mjs';
import '../shared/better-auth.BRFtaovt.mjs';
import '@better-auth/utils/binary';
import '../shared/better-auth.D2xndZ2p.mjs';
import '@better-auth/utils/hash';
import '../crypto/index.mjs';
import '@noble/ciphers/chacha.js';
import '@noble/ciphers/utils.js';
import 'jose';
import '@noble/hashes/scrypt.js';
import '@better-auth/utils/hex';
import '@noble/hashes/utils.js';
import '../shared/better-auth.B4Qoxdgc.mjs';
import '@better-auth/utils/random';
import '@better-fetch/fetch';
import 'jose/errors';
import '../shared/better-auth.BUPPRXfK.mjs';
import '@better-auth/core/db';
import 'kysely';
import 'defu';
