// db/src/generated/index-browser.js

/* !!! This is code generated by Prisma. Do not edit directly. !!!
/* eslint-disable */

Object.defineProperty(exports, "__esModule", { value: true });

const {
  Decimal,
  objectEnumValues,
  makeStrictEnum,
  Public,
  getRuntime,
  skip
} = require('./runtime/index-browser.js')


const Prisma = {}

exports.Prisma = Prisma
exports.$Enums = {}

/**
 * Prisma Client JS version: 6.17.0
 * Query Engine version: c0aafc03b8ef6cdced8654b9a817999e02457d6a
 */
Prisma.prismaVersion = {
  client: "6.17.0",
  engine: "c0aafc03b8ef6cdced8654b9a817999e02457d6a"
}

Prisma.PrismaClientKnownRequestError = () => {
  const runtimeName = getRuntime().prettyName;
  throw new Error(`PrismaClientKnownRequestError is unable to run in this browser environment, or has been bundled for the browser (running in ${runtimeName}).
In case this error is unexpected for you, please report it in https://pris.ly/prisma-prisma-bug-report`,
)};
Prisma.PrismaClientUnknownRequestError = () => {
  const runtimeName = getRuntime().prettyName;
  throw new Error(`PrismaClientUnknownRequestError is unable to run in this browser environment, or has been bundled for the browser (running in ${runtimeName}).
In case this error is unexpected for you, please report it in https://pris.ly/prisma-prisma-bug-report`,
)}
Prisma.PrismaClientRustPanicError = () => {
  const runtimeName = getRuntime().prettyName;
  throw new Error(`PrismaClientRustPanicError is unable to run in this browser environment, or has been bundled for the browser (running in ${runtimeName}).
In case this error is unexpected for you, please report it in https://pris.ly/prisma-prisma-bug-report`,
)}
Prisma.PrismaClientInitializationError = () => {
  const runtimeName = getRuntime().prettyName;
  throw new Error(`PrismaClientInitializationError is unable to run in this browser environment, or has been bundled for the browser (running in ${runtimeName}).
In case this error is unexpected for you, please report it in https://pris.ly/prisma-prisma-bug-report`,
)}
Prisma.PrismaClientValidationError = () => {
  const runtimeName = getRuntime().prettyName;
  throw new Error(`PrismaClientValidationError is unable to run in this browser environment, or has been bundled for the browser (running in ${runtimeName}).
In case this error is unexpected for you, please report it in https://pris.ly/prisma-prisma-bug-report`,
)}
Prisma.Decimal = Decimal

/**
 * Re-export of sql-template-tag
 */
Prisma.sql = () => {
  const runtimeName = getRuntime().prettyName;
  throw new Error(`sqltag is unable to run in this browser environment, or has been bundled for the browser (running in ${runtimeName}).
In case this error is unexpected for you, please report it in https://pris.ly/prisma-prisma-bug-report`,
)}
Prisma.empty = () => {
  const runtimeName = getRuntime().prettyName;
  throw new Error(`empty is unable to run in this browser environment, or has been bundled for the browser (running in ${runtimeName}).
In case this error is unexpected for you, please report it in https://pris.ly/prisma-prisma-bug-report`,
)}
Prisma.join = () => {
  const runtimeName = getRuntime().prettyName;
  throw new Error(`join is unable to run in this browser environment, or has been bundled for the browser (running in ${runtimeName}).
In case this error is unexpected for you, please report it in https://pris.ly/prisma-prisma-bug-report`,
)}
Prisma.raw = () => {
  const runtimeName = getRuntime().prettyName;
  throw new Error(`raw is unable to run in this browser environment, or has been bundled for the browser (running in ${runtimeName}).
In case this error is unexpected for you, please report it in https://pris.ly/prisma-prisma-bug-report`,
)}
Prisma.validator = Public.validator

/**
* Extensions
*/
Prisma.getExtensionContext = () => {
  const runtimeName = getRuntime().prettyName;
  throw new Error(`Extensions.getExtensionContext is unable to run in this browser environment, or has been bundled for the browser (running in ${runtimeName}).
In case this error is unexpected for you, please report it in https://pris.ly/prisma-prisma-bug-report`,
)}
Prisma.defineExtension = () => {
  const runtimeName = getRuntime().prettyName;
  throw new Error(`Extensions.defineExtension is unable to run in this browser environment, or has been bundled for the browser (running in ${runtimeName}).
In case this error is unexpected for you, please report it in https://pris.ly/prisma-prisma-bug-report`,
)}

/**
 * Shorthand utilities for JSON filtering
 */
Prisma.DbNull = objectEnumValues.instances.DbNull
Prisma.JsonNull = objectEnumValues.instances.JsonNull
Prisma.AnyNull = objectEnumValues.instances.AnyNull

Prisma.NullTypes = {
  DbNull: objectEnumValues.classes.DbNull,
  JsonNull: objectEnumValues.classes.JsonNull,
  AnyNull: objectEnumValues.classes.AnyNull
}



/**
 * Enums
 */

exports.Prisma.TransactionIsolationLevel = makeStrictEnum({
  ReadUncommitted: 'ReadUncommitted',
  ReadCommitted: 'ReadCommitted',
  RepeatableRead: 'RepeatableRead',
  Serializable: 'Serializable'
});

exports.Prisma.GlobalConfigScalarFieldEnum = {
  id: 'id',
  active: 'active',
  demoClubId: 'demoClubId',
  demoSeriesId: 'demoSeriesId',
  demoDataId: 'demoDataId',
  demoUUID: 'demoUUID'
};

exports.Prisma.UserScalarFieldEnum = {
  id: 'id',
  username: 'username',
  uuid: 'uuid',
  startPage: 'startPage',
  admin: 'admin',
  name: 'name',
  email: 'email',
  emailVerified: 'emailVerified',
  image: 'image',
  createdAt: 'createdAt',
  updatedAt: 'updatedAt',
  displayUsername: 'displayUsername',
  clubId: 'clubId'
};

exports.Prisma.RoleScalarFieldEnum = {
  id: 'id',
  name: 'name',
  clubId: 'clubId',
  permissions: 'permissions'
};

exports.Prisma.BoatScalarFieldEnum = {
  id: 'id',
  name: 'name',
  crew: 'crew',
  py: 'py',
  pursuitStartTime: 'pursuitStartTime',
  clubId: 'clubId'
};

exports.Prisma.ClubScalarFieldEnum = {
  id: 'id',
  name: 'name',
  settings: 'settings',
  displayName: 'displayName',
  stripeCustomerId: 'stripeCustomerId',
  userId: 'userId'
};

exports.Prisma.StripeScalarFieldEnum = {
  customerId: 'customerId',
  subscriptionId: 'subscriptionId',
  productId: 'productId',
  planName: 'planName',
  subscriptionStatus: 'subscriptionStatus',
  updatedAt: 'updatedAt'
};

exports.Prisma.SeriesScalarFieldEnum = {
  id: 'id',
  name: 'name',
  clubId: 'clubId',
  settings: 'settings'
};

exports.Prisma.StartSequenceScalarFieldEnum = {
  id: 'id',
  seriesId: 'seriesId',
  time: 'time',
  name: 'name',
  hoot: 'hoot',
  fleetStart: 'fleetStart',
  flagStatus: 'flagStatus',
  order: 'order'
};

exports.Prisma.RaceScalarFieldEnum = {
  id: 'id',
  number: 'number',
  Time: 'Time',
  Type: 'Type',
  seriesId: 'seriesId',
  Duties: 'Duties'
};

exports.Prisma.FleetSettingsScalarFieldEnum = {
  id: 'id',
  seriesId: 'seriesId',
  name: 'name'
};

exports.Prisma.FleetScalarFieldEnum = {
  id: 'id',
  raceId: 'raceId',
  startTime: 'startTime',
  settingsId: 'settingsId'
};

exports.Prisma.ResultScalarFieldEnum = {
  id: 'id',
  fleetId: 'fleetId',
  Helm: 'Helm',
  Crew: 'Crew',
  boatId: 'boatId',
  SailNumber: 'SailNumber',
  finishTime: 'finishTime',
  CorrectedTime: 'CorrectedTime',
  HandicapPosition: 'HandicapPosition',
  PursuitPosition: 'PursuitPosition',
  isDeleted: 'isDeleted',
  resultCode: 'resultCode',
  numberLaps: 'numberLaps'
};

exports.Prisma.LapScalarFieldEnum = {
  id: 'id',
  resultId: 'resultId',
  time: 'time',
  isDeleted: 'isDeleted'
};

exports.Prisma.TrackerScalarFieldEnum = {
  trackerID: 'trackerID',
  clubId: 'clubId',
  name: 'name'
};

exports.Prisma.BuoyScalarFieldEnum = {
  id: 'id',
  name: 'name',
  clubId: 'clubId',
  pos: 'pos',
  isMoveable: 'isMoveable',
  trackerID: 'trackerID'
};

exports.Prisma.SessionScalarFieldEnum = {
  id: 'id',
  expiresAt: 'expiresAt',
  token: 'token',
  createdAt: 'createdAt',
  updatedAt: 'updatedAt',
  ipAddress: 'ipAddress',
  userAgent: 'userAgent',
  userId: 'userId'
};

exports.Prisma.AccountScalarFieldEnum = {
  id: 'id',
  accountId: 'accountId',
  providerId: 'providerId',
  userId: 'userId',
  accessToken: 'accessToken',
  refreshToken: 'refreshToken',
  idToken: 'idToken',
  accessTokenExpiresAt: 'accessTokenExpiresAt',
  refreshTokenExpiresAt: 'refreshTokenExpiresAt',
  scope: 'scope',
  password: 'password',
  createdAt: 'createdAt',
  updatedAt: 'updatedAt'
};

exports.Prisma.VerificationScalarFieldEnum = {
  id: 'id',
  identifier: 'identifier',
  value: 'value',
  expiresAt: 'expiresAt',
  createdAt: 'createdAt',
  updatedAt: 'updatedAt'
};

exports.Prisma.SortOrder = {
  asc: 'asc',
  desc: 'desc'
};

exports.Prisma.JsonNullValueInput = {
  JsonNull: Prisma.JsonNull
};

exports.Prisma.globalConfigOrderByRelevanceFieldEnum = {
  id: 'id',
  demoClubId: 'demoClubId',
  demoSeriesId: 'demoSeriesId',
  demoDataId: 'demoDataId',
  demoUUID: 'demoUUID'
};

exports.Prisma.NullsOrder = {
  first: 'first',
  last: 'last'
};

exports.Prisma.UserOrderByRelevanceFieldEnum = {
  id: 'id',
  username: 'username',
  uuid: 'uuid',
  startPage: 'startPage',
  name: 'name',
  email: 'email',
  image: 'image',
  displayUsername: 'displayUsername',
  clubId: 'clubId'
};

exports.Prisma.JsonNullValueFilter = {
  DbNull: Prisma.DbNull,
  JsonNull: Prisma.JsonNull,
  AnyNull: Prisma.AnyNull
};

exports.Prisma.QueryMode = {
  default: 'default',
  insensitive: 'insensitive'
};

exports.Prisma.RoleOrderByRelevanceFieldEnum = {
  id: 'id',
  name: 'name',
  clubId: 'clubId'
};

exports.Prisma.BoatOrderByRelevanceFieldEnum = {
  id: 'id',
  name: 'name',
  clubId: 'clubId'
};

exports.Prisma.ClubOrderByRelevanceFieldEnum = {
  id: 'id',
  name: 'name',
  displayName: 'displayName',
  stripeCustomerId: 'stripeCustomerId',
  userId: 'userId'
};

exports.Prisma.StripeOrderByRelevanceFieldEnum = {
  customerId: 'customerId',
  subscriptionId: 'subscriptionId',
  productId: 'productId',
  planName: 'planName',
  subscriptionStatus: 'subscriptionStatus',
  updatedAt: 'updatedAt'
};

exports.Prisma.SeriesOrderByRelevanceFieldEnum = {
  id: 'id',
  name: 'name',
  clubId: 'clubId'
};

exports.Prisma.StartSequenceOrderByRelevanceFieldEnum = {
  id: 'id',
  seriesId: 'seriesId',
  name: 'name',
  fleetStart: 'fleetStart'
};

exports.Prisma.RaceOrderByRelevanceFieldEnum = {
  id: 'id',
  Time: 'Time',
  Type: 'Type',
  seriesId: 'seriesId'
};

exports.Prisma.FleetSettingsOrderByRelevanceFieldEnum = {
  id: 'id',
  seriesId: 'seriesId',
  name: 'name'
};

exports.Prisma.FleetOrderByRelevanceFieldEnum = {
  id: 'id',
  raceId: 'raceId',
  settingsId: 'settingsId'
};

exports.Prisma.ResultOrderByRelevanceFieldEnum = {
  id: 'id',
  fleetId: 'fleetId',
  Helm: 'Helm',
  Crew: 'Crew',
  boatId: 'boatId',
  SailNumber: 'SailNumber',
  resultCode: 'resultCode'
};

exports.Prisma.LapOrderByRelevanceFieldEnum = {
  id: 'id',
  resultId: 'resultId'
};

exports.Prisma.TrackerOrderByRelevanceFieldEnum = {
  trackerID: 'trackerID',
  clubId: 'clubId',
  name: 'name'
};

exports.Prisma.BuoyOrderByRelevanceFieldEnum = {
  id: 'id',
  name: 'name',
  clubId: 'clubId',
  trackerID: 'trackerID'
};

exports.Prisma.SessionOrderByRelevanceFieldEnum = {
  id: 'id',
  token: 'token',
  ipAddress: 'ipAddress',
  userAgent: 'userAgent',
  userId: 'userId'
};

exports.Prisma.AccountOrderByRelevanceFieldEnum = {
  id: 'id',
  accountId: 'accountId',
  providerId: 'providerId',
  userId: 'userId',
  accessToken: 'accessToken',
  refreshToken: 'refreshToken',
  idToken: 'idToken',
  scope: 'scope',
  password: 'password'
};

exports.Prisma.VerificationOrderByRelevanceFieldEnum = {
  id: 'id',
  identifier: 'identifier',
  value: 'value'
};


exports.Prisma.ModelName = {
  globalConfig: 'globalConfig',
  User: 'User',
  Role: 'Role',
  Boat: 'Boat',
  Club: 'Club',
  Stripe: 'Stripe',
  Series: 'Series',
  StartSequence: 'StartSequence',
  Race: 'Race',
  FleetSettings: 'FleetSettings',
  Fleet: 'Fleet',
  Result: 'Result',
  Lap: 'Lap',
  Tracker: 'Tracker',
  Buoy: 'Buoy',
  Session: 'Session',
  Account: 'Account',
  Verification: 'Verification'
};

/**
 * This is a stub Prisma Client that will error at runtime if called.
 */
class PrismaClient {
  constructor() {
    return new Proxy(this, {
      get(target, prop) {
        let message
        const runtime = getRuntime()
        if (runtime.isEdge) {
          message = `PrismaClient is not configured to run in ${runtime.prettyName}. In order to run Prisma Client on edge runtime, either:
- Use Prisma Accelerate: https://pris.ly/d/accelerate
- Use Driver Adapters: https://pris.ly/d/driver-adapters
`;
        } else {
          message = 'PrismaClient is unable to run in this browser environment, or has been bundled for the browser (running in `' + runtime.prettyName + '`).'
        }

        message += `
If this is unexpected, please open an issue: https://pris.ly/prisma-prisma-bug-report`

        throw new Error(message)
      }
    })
  }
}

exports.PrismaClient = PrismaClient

Object.assign(exports, Prisma)
